src/compilers/v2/compiler_v2.cpp

```diff
@@ -0,0 +1,518 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#define NOMINMAX
+
+#include "clc_compiler.h"
+#include "compiler.hpp"
+#include "cache.hpp"
+#include "platform.hpp"
+
+template <typename T>
+struct unique_object : public T

```

Comment:
This is very general purpose, maybe belongs in some kind of utils header?

Reply:
IMO it's not *quite* general-purpose enough to warrant that. It requires a function-pointer destructor be provided. Normally if you have something you want to RAII, you'd define a dedicated class with dedicated cleanup - this only abstracts the fact that the cleanup in all instances will be provided by a function pointer.

If I end up wanting to use it in more places then I'd move it before copy/pasting, but I prefer to keep things scoped by default rather than let leak into other code in the project.